### ISSUE TITLE
User feedback: `kctrl` package authoring commands

### ISSUE THREAD

**Issue by rohitagg2020** [carvel-dev/kapp-controller#831]
We have created this issue to capture user feedback for kctrl alpha release.
Please add your feedback as **comments** and we will incorporate it in the description below after triaging it.

Suggestion: it will be helpful if you can add [CLI] or [Doc] to your comments/issues

************************* **Accepted feedback** ************************
- [x] [Doc] kctrl pkg release is missing in case of cert-manager example [here](https://carvel.dev/kapp-controller/docs/develop/kctrl-package-authoring/#releasing-packages)  - vmware-tanzu/carvel#523
- [x] [docs] The kctrl package init command is not visible in the screenshot right after ” If we want to package cert-manager v1.9.0, we interact with package init somewhat like this:”.  - vmware-tanzu/carvel#523
- [x] [docs] The links to the “relevant faqs” return 404.  - vmware-tanzu/carvel#523
- [ ]  [logs] When authoring and iterating or consuming packages, I would like to see logs of the contained pods. I know this is not directly related to this particular kctrl feature. I have also mentioned it https://github.com/vmware-tanzu/carvel-kapp/issues/582#issuecomment-1220484333. Maybe this is already possible 🤔? - **Create a separate issue**
- [x] [Feature Request] For local directory flow, based on whether it is a helm chart or not, `pkg init` should add `helmTemplate` in the template section. - Moved to #831
- [x] [Feature Request] An Easy way to copy same Imgpkg Bundle to multiple registries. - Tracked in #918
- [x] [Feature Request] imgpkg bundle created should be tagged with the version rather than `build-timestamp` Moved to #873 
- [x] [Feature Request] Package-resources.yml file should have human readable version rather than sha-sum. May be the shasum can be added as annotation to the file. Moved to #874

**Comment by renuy** [comment]
[Doc] `kctrl pkg release ` is missing in case of cert-manager example [here](https://carvel.dev/kapp-controller/docs/develop/kctrl-package-authoring/#releasing-packages)

**Comment by mamachanko** [comment]
[docs] The `kctrl package init` command is not visible in the screenshot right after _” If we want to package cert-manager v1.9.0, we interact with package init somewhat like this:”_.

**Comment by mamachanko** [comment]
[docs] The links to the “relevant faqs” return 404.

**Comment by mamachanko** [comment]
[logs] When authoring and iterating or consuming packages, I would like to see logs of the contained pods. I know this is not directly related to this particular kctrl feature. I have also mentioned it [over here](https://github.com/vmware-tanzu/carvel-kapp/issues/582#issuecomment-1220484333). Maybe this is already possible 🤔?

**Comment by 100mik** [comment]
To make things easier, I think we can react to resolved comments with a 🎉

**Comment by mwright-pivotal** [comment]
In some cases we find the need to first fetch an existing helm chart to tar file first locally so that we can make additional changes for testing.  For example Nvidia provides a set of sample apps https://catalog.ngc.nvidia.com/orgs/nvidia/helm-charts/video-analytics-demo. .. In the instructions we are asked to fetch the helm chart locally, then deploy.   

It would be great if kctrl package init could also allow for converting a helm chart from a local file tar file in addition to helm chart repository or git repository.   There could be some air-gapped cases where this may be required as well.

**Comment by mamachanko** [comment]
[cli] `kctrl package init` adds `creationTimestamp: null` to all the resource it creates in files. It's not a problem, but a minor distraction.

E.g.
```yaml
apiVersion: kctrl.carvel.dev/v1alpha1
kind: PackageBuild
metadata:
  creationTimestamp: null
  name: upgrades.sso.apps.tanzu.vmware.com
spec:
  # ...
```

**Comment by mamachanko** [comment]
[cli]

`kctrl package release` fails to expand `~` as home:

```shell
❯ kctrl package release -v 0.1.0 --repo-output packages

Prerequisites
1. The host must be authorized to push images to a registry (can be set up by
running `docker login`)
2. Package can be released only after `kctrl package init` has been run
successfully.

The bundle created needs to be pushed to an OCI registry. Registry URL format:
<REGISTRY_URL/REPOSITORY_NAME> e.g. index.docker.io/k8slt/sample-bundle
> Enter the registry URL (): *****

Target cluster 'https://127.0.0.1:52702' (nodes: kind-control-plane)

kbld builds images when necessary and ensures that all image references are
resolved to an immutable reference
Building images and resolving references
            | $ ytt -f /var/folders/0r/_173x1jn2d1bp7h4twh90hc80000gp/T/kapp-controller-fetch-template-deploy2470784227/0/~/workspace/upgrade-test

kctrl: Error: lstat ~/workspace/upgrade-test: no such file or directory

❯ file ~/workspace/upgrade-test
/Users/mamachanko/workspace/upgrade-test: directory

```

**Comment by mamachanko** [comment]
[cli]

`kctrl package init` suggests to follow up with `kctrl dev`, but that doesn't work ootb:

```shell
❯ kctrl package init

Welcome! Before we start, do install the latest Carvel suite of tools,
specifically ytt, imgpkg, vendir and kbld as these will be used by kctrl.

Basic Information
A package reference name must be at least three '.' separated segments,
e.g. samplepackage.corp.com
> Enter the package reference name (samplepackage.corp.com): upgrades.mamachanko.com

...

Output
Successfully updated package-build.yml
Successfully updated package-resources.yml

Next steps
Created files can be consumed in following ways:
1. Optionally, use 'kctrl dev' to deploy and test the package.
2. Use 'kctrl pkg release' to release the package.
3. Use 'kctrl pkg release --repo-output repo/' to release the package and add it
to the package repository directory.

Succeeded

❯ kctrl dev
kctrl: Error: Reading App CR configuration files:
  Expected to find at least one App or PackageInstall, but found none
```

**Comment by mamachanko** [comment]
[cli]

`kctrl pkg release` after a successful `kctrl pkg init` with local directory fails:

```shell
❯ kctrl pkg release

Prerequisites
1. The host must be authorized to push images to a registry (can be set up by
running `docker login`)
2. Package can be released only after `kctrl package init` has been run
successfully.

The bundle created needs to be pushed to an OCI registry. Registry URL format:
<REGISTRY_URL/REPOSITORY_NAME> e.g. index.docker.io/k8slt/sample-bundle
> Enter the registry URL (): ***

Target cluster 'https://127.0.0.1:52702' (nodes: kind-control-plane)

kbld builds images when necessary and ensures that all image references are
resolved to an immutable reference
Building images and resolving references
            | $ ytt -f /var/folders/0r/_173x1jn2d1bp7h4twh90hc80000gp/T/kapp-controller-fetch-template-deploy3210492922/0/Users/mbrauer/workspace/upgrade-test/manifests

kctrl: Error: lstat .imgpkg/images.yml: no such file or directory

```

**Comment by 100mik** [comment]
@mamachanko the issues with the error `lstat .imgpkg/images.yml: no such file or directory` seem to be due to inappropriate error handling, this was resolved in #842 

Where as the command `kctrl dev` is missing a `-f` flag. I think we could improve the error in the case where there are none of these required flags supplied. Something like "Expected at least 1 --file flag"

**Comment by ThomasVitale** [comment]
The new `kctrl` functionality is really great, thanks for that! I find it quite straightforward to use to author packages and repositories. However, I feel a bit lost regarding the folder structure.

It would be great if we could define some guidelines on how to structure projects for Carvel packages and Carvel repositories and document them (just like Helm charts have a well-defined, standard structure). After that, the `kctrl` package authoring experience becomes more natural and intuitive. Perhaps some samples covering a few different scenarios could help (local files, GitHub release, Git repo, Helm chart...).

Looking at some existing Carvel packages (for example, some of those used by Tanzu Community Edition), I tried to identify some common practices and used them in my own packages, but they don't match with the outcome of the new `kctrl` functionality. 

In particular, I'd be interested in knowing more about the following:

- folder structure for upstream manifests and overlays (e.g. two sibling folders `upstream` and `overlays` within a `config` folder are quite common in the TCE packages);
- folder structure for additional manifests;
- location of `Package` and `PackageMetadata` manifests, especially when the Carvel package repo and the Carvel package repository repo are distinct (does it make sense to duplicate them in both places and does the CLI provide support for this scenario?);
- does the CLI support generating the `.imgpkg/images.yml` file for the package as well?
- how is the `package-resources.yml` file supposed to evolve after the package initialization?

I'm available to provide more information as well as help with contributions.

Thanks again for the great tool!

**Comment by 100mik** [comment]
Hey @ThomasVitale ! Sorry this fell off my radar. To answer a few:
> older structure for upstream manifests and overlays (e.g. two sibling folders upstream and overlays within a config folder are quite common in the TCE packages);

`kctrl` pulls in any upstream config that needed into the `upstream` folder. And yes the recommended mode of adding additional overlays is described over [here ](https://carvel.dev/kapp-controller/docs/v0.41.0/kctrl-faq/#how-can-we-add-ytt-overlays-or-additional-configuration-to-upstream-release-artifacts).

>folder structure for additional manifests

 The documentation does not however, describe a convention for additional manifests, but those would definitely live in a folder separate from the `upstream` folder. As they will get overwritten when we run `vendir  sync` to fetch newer versions of the upstream artefact. Worth noting that any additional folders will have to be mentioned in `package-build.yml`, this can be done by running `package init` to change up original config or by directly editing the resource.

> location of Package and PackageMetadata manifests

On releasing these are created in the `./carvel-artifacts` repository. Once we have them we see two paths,
1. Publish the metadata and package as artefacts directly as a release artefact which can be applied without a repository
2. Point the `package release` command towards the directory where the `PackageRepository` repo files reside using the `--repo-output` flag as elaborated on [here](https://carvel.dev/kapp-controller/docs/v0.41.0/kctrl-package-authoring/#creating-a-package-repository), which can then be published as a repository using the `package repo release` command (also elaborated on in the same doc section). **This is the recommended path**

> does the CLI support generating the .imgpkg/images.yml file for the package as well?

Yes! `kctrl` consumes `kbld` to do this within the release command.

> how is the package-resources.yml file supposed to evolve after the package initialization?

The intent of this file is,
1. To be used by the `dev` command (as described [here](https://carvel.dev/kapp-controller/docs/v0.41.0/kctrl-package-authoring/#testing-packages-locally))
2. To allow folks to add information to the `PackageMetadata` and other `Package` fields (such as `spec.license`). IF these changes are made in `package-resources.yml` they will get copied over while releasing.

We expect that while iterating over the Package, authors will use `kctrl dev` while making changes to the Package in `package-resources.yml` once they are satisfied with it is state. They can propagate the changes to `package-build.yml` before releasing the Package!

**Comment by renuy** [comment]
closing this issue as separate issues have been created for one's mentioned here
